Can't define Default method for kinds watched by multiple controllers

My manager installs multiple controllers for the same kind. If I define a `Default()` method for that kind, I get an error on startup:

```
2019-06-10T16:42:05.355-0700	INFO	controller-runtime.builder	Registering a mutating webhook	{"GVK": "my-group/v1alpha1, Kind=MyKind", "path": "/mutate-my-group-v1-alpha1-my-kind"}
panic: can't register duplicate path: /mutate-my-group-v1-alpha1-my-kind

goroutine 1 [running]:
sigs.k8s.io/controller-runtime/pkg/webhook.(*Server).Register(0xc0004510e0, 0xc0000e83c0, 0x41, 0x16f2dc0, 0xc0003f5110)
	/home/ekuefler/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.0/pkg/webhook/server.go:96 +0x25a
sigs.k8s.io/controller-runtime/pkg/builder.(*Builder).doWebhook(0xc000411300, 0x16f1360, 0xc0003f50b0)
	/home/ekuefler/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.0/pkg/builder/build.go:274 +0x89c
sigs.k8s.io/controller-runtime/pkg/builder.(*Builder).Build(0xc000411300, 0x16f1360, 0xc0003f50b0, 0x0, 0x1, 0xc000495180, 0x0)
	/home/ekuefler/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.0/pkg/builder/build.go:164 +0xcc
sigs.k8s.io/controller-runtime/pkg/builder.(*Builder).Complete(...)
	/home/ekuefler/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.0/pkg/builder/build.go:136
environments/controllers/localhost.(*ConfigMapReconciler).setupWithManager(0xc0003f50b0, 0x17478c0, 0xc0002fb680, 0x172ce60, 0xc0004aaf60)
	/home/ekuefler/environments/controllers/localhost/config_map_reconciler.go:91 +0xfd
environments/controllers/localhost.(*Reconciler).SetupWithManager(0xc000127ea8, 0x17478c0, 0xc0002fb680, 0x172ce60, 0xc0004aab40)
	/home/ekuefler/environments/controllers/localhost/reconciler.go:32 +0x1b8
main.main()
	/home/ekuefler/environments/main.go:55 +0x330
```

Presumably this is due to https://github.com/kubernetes-sigs/controller-runtime/blob/056a18adfb38c7cafac2aeefdda0b63c41c0919b/pkg/builder/build.go#L287, where the builder tries to register a new path in the manager every time a controller is created.

I may be doing something wrong, as the webhook stuff seems mostly undocumented and I'm just trying to figure it out by picking through code.

--- a/pkg/builder/build.go
+++ b/pkg/builder/build.go
@@ -18,6 +18,8 @@ package builder
 
 import (
 	"fmt"
+	"net/http"
+	"net/url"
 	"strings"
 
 	"k8s.io/apimachinery/pkg/runtime"
@@ -280,11 +282,15 @@ func (blder *Builder) doWebhook() error {
 		mwh := admission.DefaultingWebhookFor(defaulter)
 		if mwh != nil {
 			path := generateMutatePath(gvk)
-			log.Info("Registering a mutating webhook",
-				"GVK", gvk,
-				"path", path)
 
-			blder.mgr.GetWebhookServer().Register(path, mwh)
+			// Checking if the path is already registered.
+			// If so, just skip it.
+			if !blder.isAlreadyHandled(path) {
+				log.Info("Registering a mutating webhook",
+					"GVK", gvk,
+					"path", path)
+				blder.mgr.GetWebhookServer().Register(path, mwh)
+			}
 		}
 	}
 
@@ -292,10 +298,15 @@ func (blder *Builder) doWebhook() error {
 		vwh := admission.ValidatingWebhookFor(validator)
 		if vwh != nil {
 			path := generateValidatePath(gvk)
-			log.Info("Registering a validating webhook",
-				"GVK", gvk,
-				"path", path)
-			blder.mgr.GetWebhookServer().Register(path, vwh)
+
+			// Checking if the path is already registered.
+			// If so, just skip it.
+			if !blder.isAlreadyHandled(path) {
+				log.Info("Registering a validating webhook",
+					"GVK", gvk,
+					"path", path)
+				blder.mgr.GetWebhookServer().Register(path, vwh)
+			}
 		}
 	}
 
@@ -304,6 +315,14 @@ func (blder *Builder) doWebhook() error {
 		log.Error(err, "conversion check failed", "GVK", gvk)
 	}
 	return nil
+}
+
+func (blder *Builder) isAlreadyHandled(path string) bool {
+	h, p := blder.mgr.GetWebhookServer().WebhookMux.Handler(&http.Request{URL: &url.URL{Path: path}})
+	if p == path && h != nil {
+		return true
+	}
+	return false
 }
 
 func generateMutatePath(gvk schema.GroupVersionKind) string {

--- a/pkg/builder/build_test.go
+++ b/pkg/builder/build_test.go
@@ -350,6 +350,34 @@ var _ = Describe("application", func() {
 			Expect(w.Body).To(ContainSubstring(`"allowed":true`))
 			Expect(w.Body).To(ContainSubstring(`"code":200`))
 		})
+
+		It("should allow multiple controllers for the same kind", func() {
+			By("creating a controller manager")
+			m, err := manager.New(cfg, manager.Options{})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("registering the type in the Scheme")
+			builder := scheme.Builder{GroupVersion: testDefaultValidatorGVK.GroupVersion()}
+			builder.Register(&TestDefaultValidator{}, &TestDefaultValidatorList{})
+			err = builder.AddToScheme(m.GetScheme())
+			Expect(err).NotTo(HaveOccurred())
+
+			By("creating the 1st controller")
+			ctrl1, err := ControllerManagedBy(m).
+				For(&TestDefaultValidator{}).
+				Owns(&appsv1.ReplicaSet{}).
+				Build(noop)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ctrl1).NotTo(BeNil())
+
+			By("creating the 2nd controller")
+			ctrl2, err := ControllerManagedBy(m).
+				For(&TestDefaultValidator{}).
+				Owns(&appsv1.ReplicaSet{}).
+				Build(noop)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ctrl2).NotTo(BeNil())
+		})
 	})
 
 	Describe("Start with SimpleController", func() {

